awsagentprovider/src/main/java/software/amazon/opentelemetry/javaagent/providers/AwsMetricAttributeGenerator.java

```diff
@@ -392,6 +417,97 @@ private static void setRemoteResourceTypeAndIdentifier(SpanData span, Attributes
     }
   }
 
+  /**
+   * RemoteResourceIdentifier is populated with rule <code>
+   *     ^[db.name|]?{address} [|{port}]?
+   * </code>
+   *
+   * <pre>
+   * {address} attribute is retrieved in priority order:
+   * - {@link SemanticAttributes#SERVER_ADDRESS},
+   * - {@link SemanticAttributes#NET_PEER_NAME},
+   * - {@link SemanticAttributes#SERVER_SOCKET_ADDRESS}
+   * - {@link SemanticAttributes#DB_CONNECTION_STRING}-Hostname
+   * </pre>
+   *
+   * <pre>
+   * {port} attribute is retrieved in priority order:
+   * - {@link SemanticAttributes#SERVER_PORT},
+   * - {@link SemanticAttributes#NET_PEER_PORT},
+   * - {@link SemanticAttributes#SERVER_SOCKET_PORT}
+   * - {@link SemanticAttributes#DB_CONNECTION_STRING}-Port
+   * </pre>
+   *
+   * If address is not present, neither RemoteResourceType nor RemoteResourceIdentifier will be
+   * provided. If any attribute contains | or ^ , they will be replaced with ^| or ^^
+   */
+  private static Optional<String> getDBResourceIdentifier(SpanData span) {
+    String dbName = span.getAttributes().get(DB_NAME);
+
+    String serverAddress = span.getAttributes().get(SERVER_ADDRESS);
+    Long serverPort = span.getAttributes().get(SERVER_PORT);
+
+    String networkPeerAddress = span.getAttributes().get(NET_PEER_NAME);
+    Long networkPeerPort = span.getAttributes().get(NET_PEER_PORT);
+
+    String serverSocketAddress = span.getAttributes().get(SERVER_SOCKET_ADDRESS);
+    Long serverSocketPort = span.getAttributes().get(SERVER_SOCKET_PORT);
+
+    String connectionString = span.getAttributes().get(DB_CONNECTION_STRING);
+
+    Optional<String> dbServerAddress = Optional.empty();
+
+    if (serverAddress != null) {
+      dbServerAddress = buildDBResourceIdentifier(serverAddress, serverPort);
+    } else if (networkPeerAddress != null) {
+      dbServerAddress = buildDBResourceIdentifier(networkPeerAddress, networkPeerPort);
+    } else if (serverSocketAddress != null) {
+      dbServerAddress = buildDBResourceIdentifier(serverSocketAddress, serverSocketPort);
+    } else if (connectionString != null) {
+      dbServerAddress = buildDBResourceIdentifier(connectionString);
+    }
+
+    // return empty resource identifier if db server is not found
+    if (!dbServerAddress.isPresent()) {
+      return Optional.empty();
+    }
+
+    if (dbName != null) {
+      return Optional.of(replaceSpecialCharacters(dbName) + "|" + dbServerAddress.get());
+    }
```

Comment:
Not blocking, IMO a bit simpler/more clear:
```suggestion
    if (dbServerAddress.isPresent() && dbName != null) {
      dbServerAddress = Optional.of(replaceSpecialCharacters(dbName) + "|" + dbServerAddress.get());
    }
```